cmd/copy.go

```diff
@@ -1136,18 +1138,50 @@ func (cca *cookedCopyCmdArgs) processRedirectionDownload(blobResource common.Res
 	if cca.cpkOptions.IsSourceEncrypted {
 		clientProvidedKey = common.GetClientProvidedKey(cca.cpkOptions)
 	}
-	blobStream, err := blobURL.Download(ctx, 0, azblob.CountToEnd, azblob.BlobAccessConditions{}, false, clientProvidedKey)
+
+	// Calculate size of source blob
+	var blobSize int64
+	props, err := blobURL.GetProperties(ctx, azblob.BlobAccessConditions{}, clientProvidedKey)
 	if err != nil {
-		return fmt.Errorf("fatal: cannot download blob due to error: %s", err.Error())
+		blobSize = 0 // default
 	}
+	blobSize = props.ContentLength()
 
-	blobBody := blobStream.Body(azblob.RetryReaderOptions{MaxRetryRequests: ste.MaxRetryPerDownloadBody})
-	defer blobBody.Close()
+	if blobSize > 0 {
+		if cca.blockSize == 0 {
+			cca.blockSize = pipingDefaultBlockSize
+		}
 
-	// step 4: pipe everything into Stdout
-	_, err = io.Copy(os.Stdout, blobBody)
-	if err != nil {
-		return fmt.Errorf("fatal: cannot download blob to Stdout due to error: %s", err.Error())
+		// Extracted conurrency value from AZCOPY_CONCURRENCY_VALUE
+		concurrencyValue, err := strconv.Atoi(glcm.GetEnvironmentVariable(common.EEnvironmentVariable.ConcurrencyValue()))
+		if err != nil {
+			concurrencyValue = pipingDownloadParallelism
+		}
+
+		// Prepare and do parallel download.
+		err = azblob.DoBatchTransfer(ctx, azblob.BatchTransferOptions{
+			TransferSize:  blobSize,
+			ChunkSize:     cca.blockSize,
+			Parallelism:   uint16(concurrencyValue),
+			Operation: func(chunkStart int64, count int64, ctx context.Context) error {
+				dr, err := blobURL.Download(ctx, chunkStart, count, azblob.BlobAccessConditions{}, false, clientProvidedKey)
+				if err != nil {
+					return err
+				}
+				body := dr.Body(azblob.RetryReaderOptions{MaxRetryRequests: ste.MaxRetryPerDownloadBody})
+				// step 4: pipe everything into Stdout
+				_, err = io.Copy(os.Stdout, body)
```

Comment:
Wait, I don't think this is going to work. If each chunk is just writing to the Stdout whenever they are done, doesn't that corrupt the file content?

Reply:
I think we need a synchronized writer, to write all the chunks out serially (they can be downloaded in parallel).